torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1050,57 +1097,70 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
   std::lock_guard<std::mutex> lock(mutex_);
   abortCommsFromMap(devNCCLCommMap_, abortReason);
   abortCommsFromMap(inInitializationCommMap_, abortReason);
+  return true;
 }
 
 void ProcessGroupNCCL::shutdown() {
   // Don't join threads here since the purpose of this method is to abort all
   // communicators and signal the threads to exit. Joining on the threads could
   // potentially block and hence avoid it in this method.
   terminateProcessGroup_.store(true);
+  workMetaListCV_.notify_one();
 
   std::string abortReason = c10::str("Process Group shutdown on rank ", rank_);
-  abort(abortReason);
+  // lauch abort asynchrounously and wait for it to complete or timeout
+  LOG(INFO) << logPrefix()
+            << "Launching ProcessGroupNCCL abort asynchrounously.";
+  std::future<bool> fut = std::async(std::launch::async, [this, abortReason]() {
+    return this->abort(abortReason);
+  });
 
-  workMetaListCV_.notify_one();
+  waitForFutureOrTimeout(fut, options_->timeout, "ProcessGroup abort");
```

Comment:
One more thing:
If we had to use future on ncclCommAbort, we might as well just consider moving back to ncclCommDestroy.

Previously we prefer ncclCommAbort over ncclCommDestroy during regular teardown because it will not hang. Now that this characteristic is gone, and ncclCommAbort is not waitable (the only NCCL API not supporting nonblocking mode because it regards itself not possible to hang, lol), we can go back to ncclCommDestroy, but this time with nonblocking mode (i.e. in case ncclCommDestroy times out, kill it).

Reply:
Good point, I also considered using nonblocking mode (but did not know that ncclCommAbort not supporting it and we need to change to ncclCommDestroy). Since nonblocking mode rollout has wider impact other than just abort() but also initialiation and every nccl call, we'd better have a separate rollout plan on that, during which we can try change to ncclCommDestroy